### PR TITLE
Added personalisation var limit

### PR DESCRIPTION
**Pull request by jimleroyer**
# Summary | Résumé

## Size limit on the overall template variables size

Added a size limit on the overall template variables size in order to avoid notifications stuck in SQS queues, which has a limit of 256k. This can be configured via the `PERSONALISATION_SIZE_LIMIT` environment variable. The default is set to `50kb`. The limit was determined by looking at our historic data of the past week and I found the max personalisation data we got was around 17kb.

This could be even too low considering that a batch of 10 notifications can be queued in SQS and if each of these have a size of 40kb, then only 7 are necessary to overflow the SQS limit.

## Limit on the number of attachments

Added a limit on the number of attachments a notification can contain. This can be configured via the `ATTACHMENT_NUM_LIMIT` environment variable. The default is set to `10`.

# Test instructions | Instructions pour tester la modification

* Added new automated tests that test the conditional edges on number of attachments and variable overall size.

# Release Instructions | Instructions pour le déploiement

Let's check in the history of notifications of the past week...

* ...how many notifications contained more than 10 attachments.
* ...how many notifications got its personalisation data past the 50kb size.

# Reviewer checklist | Liste de vérification du réviseur

This is a suggested checklist of questions reviewers might ask during their
review | Voici une suggestion de liste de vérification comprenant des questions
que les réviseurs pourraient poser pendant leur examen :

- [ ] Is the code maintainable? | Est-ce que le code peut être maintenu?
- [ ] Have you tested it? | L’avez-vous testé?
- [ ] Are there automated tests? | Y a-t-il des tests automatisés?
- [ ] Does this cause automated test coverage to drop? | Est-ce que ça entraîne
      une baisse de la quantité de code couvert par les tests automatisés?
- [ ] Does this break existing functionality? | Est-ce que ça brise une
      fonctionnalité existante?
- [ ] Does this change the privacy policy? | Est-ce que ça entraîne une
      modification de la politique de confidentialité?
- [ ] Does this introduce any security concerns? | Est-ce que ça introduit des
      préoccupations liées à la sécurité?
- [ ] Does this significantly alter performance? | Est-ce que ça modifie de
      façon importante la performance?
- [ ] What is the risk level of using added dependencies? | Quel est le degré de
      risque d’utiliser des dépendances ajoutées?
- [ ] Should any documentation be updated as a result of this? (i.e. README
      setup, etc.) | Faudra-t-il mettre à jour la documentation à la suite de ce
      changement (fichier README, etc.)?

### CODE REVIEW

File: app/config.py [hunk 1]
```diff
@@ -216,6 +217,9 @@ class Config(object):
     SQLALCHEMY_POOL_RECYCLE = 300
     SQLALCHEMY_ECHO = env.bool("SQLALCHEMY_ECHO", False)
     PAGE_SIZE = 50
+    PERSONALISATION_SIZE_LIMIT = env.int(
+        "PERSONALISATION_SIZE_LIMIT", 1024 * 120
```
Comment (andrewleith) [inline app/config.py:221]: You mention 50kb in the PR description but its set to 120kb here.

File: app/notifications/validators.py [hunk 2]
```diff
@@ -274,17 +275,62 @@ def check_service_letter_contact_id(service_id, letter_contact_id, notification_
...
+    errors.extend(errors_vars)
+    errors.extend(errors_num_file)
+    errors.extend(errors_files)
+    return json_personalisation, errors
+
+
+def validate_personalisation(json_personalisation):
+    errors = []
+    values = [v for _, v in json_personalisation.items() if not isinstance(v, dict)]
+    concat_values = functools.reduce(lambda v1, v2: v1 + v2, values, "")
```
Comment (sastels) [inline app/notifications/validators.py:292]: these values are almost certainly strings, I think... just wondering if we should make sure, like returning `f"{v1}{v2}"` or something.

File: app/notifications/validators.py [hunk 3]
```diff
@@ -274,17 +275,62 @@ def check_service_letter_contact_id(service_id, letter_contact_id, notification_
...
+    json_personalisation, errors_vars = validate_personalisation(json_personalisation)
+    json_personalisation, errors_num_file = validate_personalisation_num_files(json_personalisation)
+    json_personalisation, errors_files = decode_personalisation_files(json_personalisation)
+    errors.extend(errors_vars)
+    errors.extend(errors_num_file)
+    errors.extend(errors_files)
+    return json_personalisation, errors
+
+
+def validate_personalisation(json_personalisation):
```
Comment (sastels) [inline app/notifications/validators.py:289]: could make the function name more specific, ie `validate_personalisation_size()` perhaps, similar to how you did for `validate_personalisation_num_files()`

**Comment by sastels** [comment]
> looks good to me! but should there be a manifest repo change as well

If we're good with just using the defaults for now then we don't need to add the env vars  ¯\_(ツ)_/¯

File: app/config.py (hunk 1 above)
Comment (jimleroyer) [reply app/config.py:221]: Yeah it's true! I didn't push that change yet. I want to meet with the Forms team before setting on a definitive value (meeting set on Monday), as they make use of this one-var template pattern. I could get their feedback before moving on. 

I will update the PR to 120k though in the meantime.

File: app/notifications/validators.py (hunk 2 above)
Comment (jimleroyer) [reply app/notifications/validators.py:292]: Yeah these are strings, string interpolation could work (in the lambda), but I'd still need to reduce the values.

File: app/notifications/validators.py (hunk 3 above)
Comment (jimleroyer) [reply app/notifications/validators.py:289]: Oh yes, I meant to do that. Thanks!

**Comment by jimleroyer** [comment]
> > looks good to me! but should there be a manifest repo change as well

> If we're good with just using the defaults for now then we don't need to add the env vars ¯\_(ツ)_/¯

Yeah that was my thinking for now. I can certainly introduce these though as it would be faster to react if we have an incident and want to change these quickly through manifests. So, not a blocker, but on my list let's say?

File: app/notifications/validators.py (hunk 2 above)
Comment (sastels) [reply app/notifications/validators.py:292]: 💯  Was thinking along the lines of
```
    concat_values = functools.reduce(lambda v1, v2: f"{v1}{v2}", values, "")
```

**Pull request merged**